core/src/main/scala/kafka/controller/ZookeeperClient.scala

```diff
@@ -0,0 +1,202 @@
+    }
+    countDownLatch.await()
+    responseQueue.asScala.toSeq
+  }
+
+  def getState: States = sessionContext.get().getState
+
+  def waitUntilConnectedOrExpired: Boolean = inLock(isConnectedOrExpiredLock) {
+    var state = sessionContext.get().getState
+    while (!state.isConnected && state.isAlive) {
+      isConnectedOrExpiredCondition.await()
+      state = sessionContext.get().getState
+    }
+    state.isConnected
+  }
+
+  def registerStateChangeHandler(stateChangeHandler: StateChangeHandler): Unit = {
+    stateChangeHandlerOpt = Option(stateChangeHandler)
+  }
+
+  def unregisterStateChangeHandler(): Unit = {
+    stateChangeHandlerOpt = None
+  }
+
+  def registerZNodeChangeHandler(zNodeChangeHandler: ZNodeChangeHandler): Unit = {
+    zNodeChangeHandlers.put(zNodeChangeHandler.path, zNodeChangeHandler)
+    sessionContext.get().exists(zNodeChangeHandler.path, true)
+  }
+
+  def unregisterZNodeChangeHandler(path: String): Unit = {
+    zNodeChangeHandlers.remove(path)
+  }
+
+  def registerZNodeChildChangeHandler(zNodeChildChangeHandler: ZNodeChildChangeHandler): Unit = {
+    zNodeChildChangeHandlers.put(zNodeChildChangeHandler.path, zNodeChildChangeHandler)
+    sessionContext.get().getChildren(zNodeChildChangeHandler.path, true)
+  }
+
+  def unregisterZNodeChildChangeHandler(path: String): Unit = {
+    zNodeChildChangeHandlers.remove(path)
+  }
+
+  def initialize(): Unit = inLock(initializationLock) {
+    if (!zooKeeper.getState.isAlive) {
+      zNodeChangeHandlers.clear()
+      zNodeChildChangeHandlers.clear()
+      zooKeeper = new ZooKeeper(connectString, sessionTimeout, ZookeeperClientWatcher)
+    }
+    sessionContext.set(zooKeeper)
+  }
+
+  def close(): Unit = inLock(initializationLock) {
+    zNodeChangeHandlers.clear()
+    zNodeChildChangeHandlers.clear()
+    zooKeeper.close()
+  }
+
+  private object ZookeeperClientWatcher extends Watcher {
+    override def process(event: WatchedEvent): Unit = {
+      if (event.getPath == null) {
+        inLock(isConnectedOrExpiredLock) {
+          isConnectedOrExpiredCondition.signalAll()
+        }
+        stateChangeHandlerOpt.foreach(_.handleStateChange)
+      } else if (event.getType == EventType.NodeCreated) {
+        Option(zNodeChangeHandlers.remove(event.getPath)).foreach(_.handleCreation)
+      } else if (event.getType == EventType.NodeDeleted) {
+        Option(zNodeChangeHandlers.remove(event.getPath)).foreach(_.handleDeletion)
+      } else if (event.getType == EventType.NodeDataChanged) {
+        Option(zNodeChangeHandlers.remove(event.getPath)).foreach(_.handleDataChange)
+      } else if (event.getType == EventType.NodeChildrenChanged) {
+        Option(zNodeChildChangeHandlers.remove(event.getPath)).foreach(_.handleChildChange)
+      }
+    }
+  }
+}
+
+trait StateChangeHandler {
+  def handleStateChange: Unit
+}
+
+trait ZNodeChangeHandler {
+  val path: String
+  def handleCreation: Unit
+  def handleDeletion: Unit
+  def handleDataChange: Unit
```

Comment:
In ZkClient, the 3 independent events (create, delete and change) are consolidated into a single callback handleDataChange(), since in all those cases, the processing is likely going to be the same (read the current data and act on it). So, I am wondering if we should do the same thing here for simplicity.